### ISSUE TITLE
Use standoff data in segmented texts

### ISSUE THREAD

**Issue by sujato** [suttacentral/suttacentral#1056]
> It's the job that's never started as takes longest to finish.

We have discussed many times the desirability of using standoff markup for text data on SC. With our segmented texts we finally have a chance to do this.

Just to clarify the terminology, as per Schmidt, section 1.

[standoff-properties.pdf](https://github.com/suttacentral/suttacentral/files/2259114/standoff-properties.pdf)

Standoff **markup** is mostly what we are talking about here. Standoff **properties**—which allow a lossless incorporation of metadata at any granularity—are more sophisticated, and proposing such a system is Schmidt's main thesis. These come into play in the final section of this page.

## The Problem

Currently we keep our texts in either a translation format (PO) or a presentation format (HTML). Both of these involve serious compromises. Without going into too much detail, the basic problem is that such formats, being pressed into serving a purpose for which they were not designed, end up mixing various kinds of data. Instead, we should separate each kind of data cleanly and consistently using a dedicated data format, probably JSON.

The advantage of standoff markup is that it allows an unlimited set of data to be associated with the source, without cluttering up the source files. This can include such things as:

1. Multiple references from different editions, with more that can be added over time.
2. HTML or other markup.
3. Notes.
4. Variant readings
5. Anything else!

## The Idea

The data needs to preserve three things for each item:

1. The segment ID—a number that stands for the absract idea of the segment and is unique across the entire SC corpus.
3. The content—a string of utf-8 glyphs that represents a line of text, a reference number, and so on.
2. The attribute—the kind of thing that the data is; for example an English translation of the Pali by Sujato.

These things should be preserved cleanly, simply, unambiguously, and locally (i.e. you should not have to go to another file to find out who the author is.)

We can do this by making everything a JSON object.

1. The segment ID is the key of the JSON object.
3. The content is the value of the JSON object.
2. The attribute is defined in the file.

Currently in DN 1 we have a PO file that records the following information about the first segment of DN 1.

    #. </h2><p>
    #. <a class="pts-cs" id="pts-cs1.1"></a>
    #. <a class="pts-vp-pli" id="pts-vp-pli1.1"></a>
    #. <a class="sc" id="sc1"></a>
    #. <span class="evam">
    msgctxt "dn1:1.1.1"
    msgid "Evaṃ me sutaṃ—​"
    msgstr "So I have heard."

Notice that this breaks the rules on a number of fronts:

1. Different kinds of stuff is mixed: references, HTML, source text, translation text. In other segments, variant readings or comments might also be found.
2. It is verbose, including all the stuff necessary for HTML markup of references.
2. There is nothing to identify the edition or the author.

Since each segment has a `msgctxt` number, we can split this into data sets something like the following.

    {
        "ms-pli":[
            {
                "dn1:1.1.1":"Evaṃ me sutaṃ—"
            },
            {
                "dn1:1.1.2":"ekaṃ samayaṃ bhagavā"
            }
        ]
    }

The root key tells us the attribute, which is *universal* to this data set. Here it is `ms-pli`, i.e. the Mahasangiti edition of the Pali Tipitaka. The segment IDs and text strings form key/value objects. They are ordered, so form an array.

A translation can be handled similarly.

    {
        "sujato-en":[
            {
                "dn1:1.1.1":"So I have heard."
            },
            {
                "dn1:1.1.2":"At one time the Buddha"
            }
        ]
   
    }

This tells us that it is an English translation by Sujato; that the translation is of DN 1; the identity of each segment; and the content of the translation itself.

For references we can include multiple editions within one array. It would also be possible to have each edition in a separate file. Here the key is, as always, the segment ID. This matches an object which is a set of key/value pairs, each representing a specific reference system + number.

    {
        "refs":[
            {
                "dn1:1.1.1":{
                    "pts-cs":"1.1",
                    "pts-vp-pli":"1.1",
                    "msdiv":"1"
                }
            },
            {
                "dn1:1.1.2":{
                }
            }
        ]
    }

In another file can have markup. Note that this can be done better than in PO, for in PO we can only add markup at the beginning of a segment, and hence must *close* the previous html tag where applicable—see above `</h2><p>`. This creates problems; for example, at the end of a sutta, there is nowhere for the final tags to be closed. Much better to do something like this:

    {
        "html":[
            {
                "dn1:1.1.1":"<p><span class="evam">{{1stString}}</span>"
            },
            {
                "dn1:1.1.2":"{{2ndString}}</p>"
            }
        ]
   
    }

## Usage

Once the data is cleanly separated like this, we can mix and match it as we wish. Some use cases:

- For search, just take the text. Or index the refs separately.
- For online, import just the text and HTML, and bring in other data dynamically as requested.
- For Pootle, leave out refs and HTML, but include notes and variants.
- Presentation markup can be adapted for various cases. For example, different languages might handle paragraph conventions differently.
- We have already seen cases where people have used SC's data for themselves. By cleanly separating the sources, such third-party applications can flexibly use whatever data is convenient for them.
- Things we have not thought of! We set up elegant and powerful data, and the applications will suggest themselves.

## Note!!!

There are some bugs in the current pootle data, when converting to JSON we should squash them:

https://github.com/suttacentral/suttacentral/issues/1115 

## Standoff properties: do stuff inside segments

All the above assumes that the stuff we need to remove can be located outside a segment. And in most cases, this is true. The structural markup works outside segments. Refs such as page break lose a little granularity by going outside the segment, but nothing serious—in most cases. Variants and notes can work fine that way, too.

However certain things must live inside a segment. These include:

- Certain kinds of textual emphasis or styling: *all* consciousness is not-self!
- Some references would be awkward to handle outside segments. I am thinking primarily of line-breaks in the Taisho texts. We would commonly have two or more line breaks in one segment.
- Reconstructed and other text marked with text-critical markup, especially in Sanskrit texts and other manuscripts.

To handle such cases, a system called "standoff properties" has been developed using a JSON-based markup called STIL. 

https://discourse.suttacentral.net/uploads/default/original/2X/6/6056afc3c25fcf0e9b3e677c04ea4bc34b8151ab.pdf

Essentially, each inner span would be marked with three pieces of information, which represent:

3. The kind of property (`name`)
1. The number of glyphs since the start of the segment (`reloff`)
2. The length of the inner span (`len`)

Here is an example.

![stil](https://user-images.githubusercontent.com/6112010/43619593-4ab768f4-9701-11e8-803b-d4a60457a333.png)

Unfortunately, work on the basic tools for such a system is still ongoing, and I don't think it is mature yet. I would suspect that for the time being we will have to fake it a bit. Perhaps we could use a set of defined plain text glyphs to represent text-critical info, and markdown for styling. 

BTW, this looks cool:

https://github.com/dhlab-basel/Knora/pull/319

**Comment by SamaneriVimala** [comment]
I would like to have @blake-sc's opinion here also.

Personally I'm all for standoff properties but would rather do it a bit different.
The .po files are fine as they are, we can just remove the coding and other things that are in comments. After that, they are basically just a dictionary/array and can be used as such without having to convert them to json first, leaving out another step. Pootle needs .po files and why would you convert them to separate files first if you can simply get Arango DB to read it in as it is (and output a json to the frontend with that).

So I would go with a dn1.po, dn2.po, etc. and a dn.json, the latter just being an array with segment numbers linked to whatever markup and other numbers you want in there. (or alternatively a dn1.json, etc. if dn.json is too long).

**Comment by sujato** [comment]
So the question, "why would you convert to PO files"?

Because PO files are a translation format, not a data format. They are designed for a purpose, and should be used for that purpose. And that purpose doesn't exactly fit what is needed. We should treat the fundamentals on SC as *data*, and use a data format for them. The whole point of standoff is to separate out concerns, so that one JSON file = one kind of data. 

When working with PO files, there are various problems.

1. The files include bunches of metadata that is unnecessary for simply working with text.
2. Tooling is much more primitive as opposed to JSON.
3. You have the POT file for the template, and the PO files that have source and target. But there is no basic file, no single, simple source of truth that represents a translation.
4. Pootle splits the PO files into lines, which is arbitrary, and introduces extra glyphs into the text, which can, and does, cause problems outside the Pootle world.

It doesn't really resolve the basic problem, which is that *different kinds of things are mixed up in the same file*. That is ultimately the problem with HTML and PO and all these formats. When different stuff is mixed up, it always creates problems. 

As for the file names, numbering, yes, there's no real need for any particular system, we could even put every instance of the same kind of data in a single JSON file, as the segment numbers are unique in the corpus. Basically it's whatever is handy. But HTML markup is in one file, references in another, and so on. We might—and in fact probably will—end up with multiple sets of HTML and other kinds of markup. Great, they can all be there, sitting cleanly in separate files.

Also just to note I used the terms "standoff properties" and "standoff markup" incorrectly previously, I will fix this.

**Comment by firepick1** [comment]
(this is Karl)
1) How should we represent sutta paragraph/section semantics in JSON?  I think this will be critical for assisted search.
2) The JSON format should allow us to present both the full sutta view as well as the contracted ellipsis view for use in both textual and audio contexts. The ellipsis view is critical for skimming comprehension, the full view is equally valuable for the very experience of the deliberate, measured and unfolding traversal of the single Dhamma thread that is a sutta.
3) GUIDs generated from content may be useful for references in the proposed new format. For example, I am using GUIDs to track sound segments for caching to limit calls to TTS voice services. GUIDs can also be used for Merkle trees where aggregates have a GUID generated from the GUID's of each of their parts.

**Comment by sujato** [comment]
First of all, just to be clear, I am speaking here solely of the *segmented texts*, i.e. my new translations, Brahmali's new translations, and subsequent texts. Others are "legacy html" which is a deprecated format. We maintain them as older texts and will continue to add to them, but new features don't target them. We build for segmented texts, and if we have time and interest, maybe the legacy html.

1. Suttas are divided into "sections" (roughly a paragraph, but not exactly), and "segment" (a sentence or semantic unit of a sentence, eg. a clause.). In an ID like `dn1.1.1`, `dn1` is the sutta, `dn1.1` is the section, and `dn1.1.1` is the segment. The segments are what you see if you enable "line by line view".
2. There is no full view vs. contracted view, nor will there be in the forseeable future. See discussions [here](https://discourse.suttacentral.net/t/translating-the-four-nikayas/341/103) and [here](https://discourse.suttacentral.net/t/cutting-too-much-out-of-suttas/8344). In brief, it is extremely complicated and time-consuming, it does not reflect the manuscript tradition, and it has little use outside of specialized studies. An issue to be considered, though, is that the human-read versions sometimes use texts altered by the readers for reading aloud. 
3. How does this help where our IDs don't? The segment IDs are unique within our corpus.

**Comment by firepick1** [comment]
Yes. Effort should clearly go towards future segmentation. I just wanted to note in passing that tokenization, segmentation are already required for TTS, so automatic segmenting of legacy HTML might help  end users wanting to reference alternate translations of a segment.

1. Aha! Thanks for the numbering. That's perfect.
2. If an easy solution presents itself, I'll investigate given that "the very same features that are most annoying in a written text often have exactly the opposite effect in recitation."
3. The current ID system is sutta local. We can informally refer to "SuttaCentral MN1 Sujato 172-194.25Why is that?", but that would be a human convention with no formal syntax. A global identification system/scheme for specific segments inside and outside of SuttaCentral would provide unity of reference across all translations and sources.  For example a JSON description { url:"...", segment:"SC172-194.25"} would provide a canonical reference that could be used directly or converted into a corresponding GUID for implementation efficiency. I'm having to do this internally for tracking sound segments, so it might or might not be of value in a larger context.

**Comment by blake-sc** [comment]
I've been thinking a lot about this and I think it's a really exciting
idea that could add a lot of value. After giving it due reflection,
here is how I would go about it: 


```
FILE: dn/pli/dn1.json
{
  "_meta": {
    "type": "source",
    "language": "pli",
    "origin": "ms"
  },
  "dn1:1.1.1": "Evaṃ me sutaṃ—",
  "dn1:1.1.2": "ekaṃ samayaṃ bhagavā"
 }
```
Each document is of key:value pairs, where the key is an id and the
value can be a string or an object. A special "_meta" key allows
setting extra meta properties. I would also propose allowing _meta inheritance
that is the _meta field can be ommitted and it can be inherited from
a document named _meta.json containing the fields.

Note that the JSON format doesn't maintain the order of keys,
nevertheless in practise most engines do, and if we can guarantee
that the keys can be put in proper order through a sort algorithm
as is possible with our scheme it is okay to use a plain object.

An advantage of using a plain object is these can be loaded as JSON
and are immediately ready to be consumed in Javascript or Python. That
it is also the simplest possible format is a nice bonus.
 <br><br>

```
FILE: dn/en/dn1.json
{
  "_meta": {
    "type": "translation",
    "language": "en",
    "translator": "sujato"
  },
  "dn1:1.1.1": "So I have heard.",
  "dn1:1.1.2": "At one time the Buddha"
}
```
A translation is the same as a source except the contents of _meta
which declare it to be a translation.
 <br><br>

```
FILE: refs/dn/dn1.json
{
  "_meta": {
    "type": "reference"
  },
  "dn1:1.1.1":{
    "msdiv":"1",
    "pts-cs":"1.1",
    "pts-vp-pli":"1.1"
  }
}
```
If there's no data for an id no need to have anything.
Also this is an example of using an object rather than a string.
 <br><br>

```
FILE: dn/markup/dn1.json
{
  "_meta": {
    "type": "html_markup"
  },
  "dn1:1.1.1":"<p><span class="evam">{}</span>",
  "dn1:1.1.2":"{}</p>"
}
```
In Python empty squiggly brackets can be used to substitute "the first
thing".
 <br><br>

```
FILE: comments/dn/dn1.json
{
  "_meta": {
    "type": "comment",
  },
  "dn1:1.1.1": {
    "captain_obvious": "A common way that suttas start"
  }
}
```
Could also be used for comments, variant notes etc.
 <br><br>

As a general side note: The folder structure could put all things
related to one sutta together, or it could put all things related
to one language together or you could be silly and put everything into 
randomly named folders and files. The ids themselves are what bind the
documents together so it doesn't really matter what file names
and folder structures are used for the json files.

So for example we could also use files like this:  

* dn/dn1_pli_ms.json  
* dn/dn1_en_sujato.json  
* dn/dn1_markup.json  
* dn/dn1_refs.json  
* dn/dn1_comments.json  

It is the content of the files that determines what they are not their name.
 <br><br>

Also these documents could also be combined into one file:

```
[
  {
    "_meta": {
      "type": "source",
      "language": "pli",
      "origin": "ms"
    },
    "dn1:1.1.1":"Evaṃ me sutaṃ—",
    "dn1:1.1.2":"ekaṃ samayaṃ bhagavā"
  },
  {
    "_meta": {
      "type": "translation",
      "language": "en",
      "translator": "sujato"
    },
    "dn1:1.1.1":"So I have heard.",
    "dn1:1.1.2":"At one time the Buddha"
  },
  {
    "_meta": {
      "type": "html_markup"
    },
    "dn1:1.1.1":"<p><span class="evam">{}</span>",
    "dn1:1.1.2":"{}</p>"
  }
]
```
If it is desired that multiple documents be sent together or even saved
together then they can simply be put in an array.
 <br><br>

I believe this scheme maximizes simplicity while still being powerful enough.

**Comment by blake-sc** [comment]
### Standoff Editor

Using standoff data is great and all, but it's not really great and all without a good way to edit it that eliminates any possibility of syntax errors and that allows viewing things together.

For that I would propose the creation of a standoff editor, inspired by the segmented views. Implemented in Javascript, runs in a browser or browser-like environment.

* The editor is backed directly by a git repository.
* The editor retrives two or more linked JSON documents (some of which can
be blank).
* It creates a side-by-side or line-by-line view.
* Each particular document can be set as read-only or writeable. For strings
this is as setting the "contenteditable" property on elements.
* In practise it would be desirable to implement "modes", like "translation mode"
would load a source document and load (or create an empty) translation document
and set the source to read-only and the translation to editable.
* Once editing is done, the edited document is sent back to and commited into the repo.

It's worth noting the standoff editor can start extremely simple, the basic functionality is not at all complicated and othe than needing to do POST requests is far more simple than our segmented view on SC. It can be enhanced over time with more sophisticated modes.



### Web Editor: 

I would also propose a web version of the editor. It would use a "thin server" between the git repository and client-side editor.

Something like Pootle could be considered a "fat server", it loads all the texts into a database which is seperate from any underlying repository, which then requires synchonization and is error prone
and doesn't make it easy to directly edit files in the repository.

The idea of the thin-server that *git is the database*, the thin server just has the job of fetching documents from the repo and commiting newly edited documents back in, repo synchornization isn't something tacked on, it's the foundation. git handles history and can also in principle handle merges, conflicts and suggestions (as pull requests).  
The thin server could be enhanced to also manage stats, users and permissions.

This approach would scale well to tens or hundreds of thousands of documents and a gigabyte or more of content, the editor would be lightning fast as the whole documents are sent to the client, there's no need for a go-between the server and client for each and every string.

**Comment by blake-sc** [comment]
I'm sorry I don't understand the problem, or I don't understand how the example is meant to relate to the problem.

The purest expression of standoff text is just a text file with one segment per line. How those segments are put together for presentation is of no concern to the data, in fact even by segmenting we've added something which isn't there in the oral tradition, but just having all the characters run together is not going to be user-friendly. Simply using line numbers to reference a segment is unsatisfactory, it is a little brittle. The next best thing is explicitly labeling each segment, it's a bit less pure because we've added something else which isn't intrinsic to the text, but it's less brittle.

I find it objectional to add any more structure than that. While the segment ids represent an honest effort to arrange the text into logical paragraphs it shouldn't be thought that those paragraphs are an intrinsic feature of the text, they're just one possible way of adding structure to a linear recitation, we don't want the paragraph structure set in stone in the data. If it were to be proposed the ids should just be literally increasing integers I would agree in principle though in practice it is unwieldy.

The paragraph structure is properly defined in markup. So far I'm not completely sold on the markup examples in this thread. Internally we currently use a structure like this:
```
[
  "<p><span class="evam"><sc-seg id="dn1:1.1.1"></sc-seg></span> <sc-seg id="dn1:1.1.2"></sc-seg></p>",
...
]
```

Or a templated expression could be:
```
[
  "<p><span class="evam">{dn1:1.1.1}</span> {dn1:1.1.2}</p>",
  ...
]
```

To me this actually still seems like a much more logical way of defining the overall document markup, it would even allow things like reversing the order of segments like for fancy headings although that wouldn't be a recommended use!

By the way, the primary reason to use a simple-as-possible format for the text data is not to save disk or transmission space:

1. Improve signal-to-noise ratio when viewing, reviewing or editing the files manually
2. Generate diff-friendly files
3. Make the data ready-to-use. For example in the above markup case it's literally as simple as loading the json file, then doing `data[segment_id]`.

**Comment by sujato** [comment]
Okay so this all sounds grand. We should probably continue on Discourse though, conversations on Github are bad! But anyway, just a couple of points. 

- Like Blake, I'm not entirely sure what @firepick1's example is supposed to show. But I think he has touched on a genuine issue. The underlying practical problem is that since the segment numbers are *not* a simple increment they invite problematics in the future. So perhaps we should instead use a "pure" simple increment as the base ID and map it on to reference systems for display only. 
- I'm not sure why Blake's proposal ditches the array: the data is intrinsically ordered, surely that is what arrays are for?
- For a standoff editor, we want an editor that is cross-platform, web-like, tightly integrated with Git, and readily hackable. Sounds like Atom!
- I'm still not convinced that adding all the functionality of a translation editor is going to be doable, but convince me!
- Blake's proposed syntax for embedding text in HTML does indeed make better sense. Note that it is very similar to the `lit-html` syntax: `let sayHello = (name) => html`<h1>Hello ${name}</h1>`;`.

**Comment by firepick1** [comment]
I've created two posts in SuttaCentral meta dev to broaden the discussion. There should probably be more considering the vast number of ideas packed into this thread. The new posts address concerns I've bumped into regarding immediate work on visual assistance. I think I've answered my own questions, but have posted to verify my understanding of SC architecture. The one concern that I'd like particular attention to is the proposal for ditching the array--this seems a bit risky given that object keys have no guaranteed order in JSON. Arrays specify order. Object keys can only guarantee order iff their sort order exactly matches the expected linear sequence. Given that our segment ids are not entirely consistent (see post), I'm a little worried:

"I'm not sure why Blake's proposal ditches the array: the data is intrinsically ordered, surely that is what arrays are for?"

Regarding the editors I do like the idea, but I'm not a translator, am not a target user, and my opinion shouldn't really count for prioritization.

**Comment by sujato** [comment]
I'll respond over at Discourse. Note that since the last post here Blake and I have had an extensive discussion on this. The reason he favors ditching the array is that it adds complexity, and that agents consume JSON in default order anyway. But this is considering only the agents he is thinking about, so let us know if it is different in your case. Either way, though, it wouldn't be difficult to transform it by adding an array.

**Comment by sujato** [comment]
This has essentially been implemented in bilara-data.
